### PR TITLE
Add Astro forum docs

### PR DESCRIPTION
**Pull request by pankajastro**
closes: https://github.com/astronomer/ask-astro/issues/120

### CODE REVIEW

**Comment by pankajastro** [comment]
@vatsrahul1001 We need to test this PR too. please let me know when you are planning to test I'll keep ready db accordingly

File: airflow/dags/ingestion/ask-astro-load.py [hunk 1]
```diff
@@ -8,9 +8,15 @@
 
 import pandas as pd
 from include.tasks import split
+<<<<<<< HEAD
```
Comment (pankajastro) [inline airflow/dags/ingestion/ask-astro-load.py:11]:
```suggestion
```

File: airflow/dags/ingestion/ask-astro-forum-load.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+@dag(
+    schedule_interval=schedule_interval,
+    start_date=datetime.datetime(2023, 9, 27),
+    catchup=False,
+    is_paused_upon_creation=True,
+    default_args=default_args,
+)
+def ask_astro_load_astro_forum():
+    _import_data = (
```
Comment (sunank200) [inline airflow/dags/ingestion/ask-astro-forum-load.py:36]: This should ideally be in three tasks right? 

extract_forum_docs >> split_docs >> ingest_data

File: airflow/include/tasks/extract/astro_forum_docs.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
+import logging
```
Comment (sunank200) [inline airflow/include/tasks/extract/astro_forum_docs.py:1]:
```suggestion
from __future__ import annotations

import logging
```

File: airflow/include/tasks/extract/astro_forum_docs.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+import requests
+from bs4 import BeautifulSoup
+from weaviate.util import generate_uuid5
+
+cutoff_date = datetime(2022, 1, 1, tzinfo=pytz.UTC)
+
+logger = logging.getLogger("airflow.task")
+
+
+def get_questions_urls(html_content: str):
```
Comment (sunank200) [inline airflow/include/tasks/extract/astro_forum_docs.py:15]:
```suggestion
def get_questions_urls(html_content: str) -> list:
```

File: airflow/include/tasks/extract/astro_forum_docs.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    soup = BeautifulSoup(html_content, "html.parser")
+    url_list = []
+    for a_tag in soup.findAll("a", class_="title raw-link raw-topic-link"):
+        url_list.append(a_tag.attrs.get("href"))
+    return url_list
+
+
+def get_publish_date(html_content):
```
Comment (sunank200) [inline airflow/include/tasks/extract/astro_forum_docs.py:23]:
```suggestion
def get_publish_date(html_content) -> datetime:
```

File: airflow/include/tasks/extract/astro_forum_docs.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    soup = BeautifulSoup(html_content, "html.parser")
+    # TODO: use og:article:tag for tag filter
```
Comment (sunank200) [inline airflow/include/tasks/extract/astro_forum_docs.py:25]: Should we implement this todo?

File: airflow/include/tasks/extract/astro_forum_docs.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    for question_url in questions_urls:
+        html_content = requests.get(question_url).content
+        if get_publish_date(html_content) < cutoff_date:
+            continue
+        valid_urls.append(question_url)
+    return valid_urls
+
+
+def get_cutoff_questions(forum_url):
+    page_number = 0
```
Comment (sunank200) [inline airflow/include/tasks/extract/astro_forum_docs.py:42]: can you add a doc string that this method is doing?

**Review by sunank200 — COMMENTED** [review]
Lets ingest data to a separate weaviate index and test this out.

**Comment by vatsrahul1001** [comment]
> Lets ingest data to a separate weaviate index and test this out. cc: @pankajastro @vatsrahul1001

@pankajastro let me know when index is pointed to dev

File: airflow/include/tasks/extract/astro_forum_docs.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    base_url = f"{forum_url}?page="
+    all_valid_url = []
+    while True:
+        page_url = f"{base_url}{page_number}"
+        logger.info(page_url)
+        page_number = page_number + 1
+        html_content = requests.get(page_url).content
+        questions_urls = get_questions_urls(html_content)
+        if len(questions_urls) == 0:  # reached at the end of page
```
Comment (Lee-W) [inline airflow/include/tasks/extract/astro_forum_docs.py:51]:
```suggestion
        if not questions_url:  # reached at the end of page
```

File: airflow/include/tasks/extract/astro_forum_docs.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            return set(all_valid_url)
+        filter_questions_urls = filter_cutoff_questions(questions_urls)
+        all_valid_url.extend(filter_questions_urls)
+
+
+def clean_content(row_content):
```
Comment (Lee-W) [inline airflow/include/tasks/extract/astro_forum_docs.py:57]:
```suggestion
def clean_content(row_content) -> str | None:
```

File: airflow/include/tasks/extract/astro_forum_docs.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    # Remove script and style tags
+    for script_or_style in soup(["script", "style"]):
+        script_or_style.extract()
+
+    # Get text and handle whitespaces
+    text = " ".join(soup.stripped_strings)
+    return text
+
+
+def fetch_url_content(url):
```
Comment (Lee-W) [inline airflow/include/tasks/extract/astro_forum_docs.py:71]:
```suggestion
def fetch_url_content(url) -> str | None:
```

File: airflow/include/tasks/extract/astro_forum_docs.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    try:
+        response = requests.get(url)
+        response.raise_for_status()  # Raise an HTTPError for bad responses
+        return response.content
+    except requests.RequestException:
+        logger.info("Error fetching content for %s: %s", url, url)
+        return None
+
+
+def process_url(url, doc_source=""):
```
Comment (Lee-W) [inline airflow/include/tasks/extract/astro_forum_docs.py:81]:
```suggestion
def process_url(url: str, doc_source: str = "") -> dict | None:
```

File: airflow/include/tasks/extract/astro_forum_docs.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    Process a URL by fetching its content, cleaning it, and generating a unique identifier (SHA) based on the cleaned content.
+    param url (str): The URL to be processed.
+    """
+    content = fetch_url_content(url)
+    if content is not None:
+        cleaned_content = clean_content(content)
+        sha = generate_uuid5(cleaned_content)
+        return {"docSource": doc_source, "sha": sha, "content": cleaned_content, "docLink": url}
+    else:
+        return None
```
Comment (Lee-W) [inline airflow/include/tasks/extract/astro_forum_docs.py:92]:
```suggestion
```

File: airflow/include/tasks/extract/astro_forum_docs.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+
+
+def url_to_df(urls, doc_source=""):
```
Comment (Lee-W) [inline airflow/include/tasks/extract/astro_forum_docs.py:95]:
```suggestion
def url_to_df(urls: list[str], doc_source: str = "") -> pd.DataFrame:
```

File: airflow/include/tasks/extract/astro_forum_docs.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+    param urls (list): A list of URLs to be processed.
+    """
+    df_data = [process_url(url, doc_source) for url in urls]
+    df_data = [entry for entry in df_data if entry is not None]  # Remove failed entries
+    df = pd.DataFrame(df_data)
+    df = df[["docSource", "sha", "content", "docLink"]]  # Reorder columns if needed
+    return df
+
+
+def get_forum_df():
```
Comment (Lee-W) [inline airflow/include/tasks/extract/astro_forum_docs.py:107]:
```suggestion
def get_forum_df() -> list[pd.DataFrame]:
```

File: airflow/include/tasks/extract/astro_forum_docs.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    questions_link = get_cutoff_questions("https://forum.astronomer.io/latest")
```
Comment (Lee-W) [inline airflow/include/tasks/extract/astro_forum_docs.py:108]:
```suggestion
    questions_links = get_cutoff_questions("https://forum.astronomer.io/latest")
```

File: airflow/include/tasks/extract/astro_forum_docs.py [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+    logger.info(questions_link)
+    df = url_to_df(questions_link, "astro-forum")
```
Comment (Lee-W) [inline airflow/include/tasks/extract/astro_forum_docs.py:110]:
```suggestion
    df = url_to_df(questions_links, "astro-forum")
```

File: airflow/include/tasks/extract/astro_forum_docs.py (hunk 7 above)
Comment (pankajastro) [reply airflow/include/tasks/extract/astro_forum_docs.py:45]: Added

File: airflow/include/tasks/extract/astro_forum_docs.py (hunk 6 above)
Comment (pankajastro) [reply airflow/include/tasks/extract/astro_forum_docs.py:24]: Currently we have just around 250docs so we can keep everything for now

File: airflow/dags/ingestion/ask-astro-forum-load.py (hunk 2 above)
Comment (pankajastro) [reply airflow/dags/ingestion/ask-astro-forum-load.py:35]: ingestion for this docs take around 1min so splitting won't add much value

Comment (pankajastro) [reply airflow/dags/ingestion/ask-astro-forum-load.py:35]: or this is requirement?

Comment (sunank200) [reply airflow/dags/ingestion/ask-astro-forum-load.py:35]: This is a reference implementation. It's good to keep different tasks broken down.

Comment (pankajastro) [reply airflow/dags/ingestion/ask-astro-forum-load.py:38]: splited it.  PTAL

**Review by sunank200 — APPROVED** [review]

**Pull request merged**